### PR TITLE
[UNR-3472] New runtime settings

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Changes to the Editor settings to be able to select a different runtime variant as well as specify their respective versions.
Took the opportunity to put the version settings in the advanced settings.

#### Folded
image

#### Unfolded

image

### CODE REVIEW

**Comment by mironec** [comment]
I guess we don't want to allow running with different types of runtimes for local and cloud?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 1]
```diff
@@ -38,12 +38,34 @@ void FSpatialLaunchConfigDescription::OnWorkerTypesChanged()
...
+
 USpatialGDKEditorSettings::USpatialGDKEditorSettings(const FObjectInitializer& ObjectInitializer)
 	: Super(ObjectInitializer)
 	, bShowSpatialServiceButton(false)
 	, bDeleteDynamicEntities(true)
 	, bGenerateDefaultLaunchConfig(true)
-	, bUseGDKPinnedRuntimeVersion(true)
+	, RuntimeVariant(ESpatialOSRuntimeVariant::Distributed)
+	, DistributedRuntimeVersion(SpatialGDKServicesConstants::SpatialOSRuntimePinnedVersion)
+	, SingleNodeRuntimeVersion(SpatialGDKServicesConstants::SpatialOSRuntimePinnedVersion)
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp:68]: This constant changes when merging with Andreas's PR yeah?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 2]
```diff
@@ -61,22 +83,14 @@ USpatialGDKEditorSettings::USpatialGDKEditorSettings(const FObjectInitializer& O
...
+FRuntimeVariantVersion& USpatialGDKEditorSettings::GetRuntimeVariantVersion(ESpatialOSRuntimeVariant::Type Variant)
 {
-	if (bUseGDKPinnedRuntimeVersion || CloudRuntimeVersion.IsEmpty())
+	if (Variant == ESpatialOSRuntimeVariant::Distributed)
 	{
-		return SpatialGDKServicesConstants::SpatialOSRuntimePinnedVersion;
+		return DistributedRuntimeVersion;
 	}
-	return CloudRuntimeVersion;
+	check(Variant == ESpatialOSRuntimeVariant::SingleNode);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp:92]: What's with the check?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialRuntimeVersionCustomization.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		// Layout other properties as usual.
+		if (ChildProperty->GetProperty()->GetFName() != PinnedGDKPropertyName)
+		{
+			StructBuilder.AddProperty(ChildProperty.ToSharedRef());
+			continue;
+		}
+
+		IDetailPropertyRow& CustomRow = StructBuilder.AddProperty(ChildProperty.ToSharedRef());
+
+		FString PinnedVersionDisplay = FString::Printf(TEXT("GDK Pinned Version : %s"), *SpatialGDKServicesConstants::SpatialOSRuntimePinnedVersion);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialRuntimeVersionCustomization.cpp:47]: This should be variant based no?

**Comment by ImprobableNic** [comment]
Well that was added because some runtime bundles had different names that needed to be used for cloud/local.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp (hunk 1 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp:68]: Yes.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp (hunk 2 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp:92]: Just to make say we do not have any other values than the two known ones.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialRuntimeVersionCustomization.cpp (hunk 3 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialRuntimeVersionCustomization.cpp:47]: Good catch!

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp (hunk 1 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp:68]: It's also been noted that Squid and Kraken will require different template names which we should update the defaults to when swithcing

**Comment by joshuahuburn** [comment]
Updated with the new agreed naming scheme. Fixed the pinned version display and updated the default versions.

**Pull request merged**